File: backend/api/chart_groups.py
"""
Chart Group API — 散点图数据点组合管理
"""
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from backend import models
from backend.database import get_db
from backend.security import get_current_user, get_current_user_optional

router = APIRouter(prefix="/api/chart-groups", tags=["图表组合"])


def _group_to_dict(g: models.ChartGroup) -> dict:
    return {
        "id": g.id,
        "name": g.name,
        "description": g.description,
        "is_preset": g.is_preset,
        "is_public": g.is_public,
        "created_by": g.created_by,
        "creator_name": g.creator.real_name if g.creator else None,
        "created_at": g.created_at.isoformat() if g.created_at else None,
        "updated_at": g.updated_at.isoformat() if g.updated_at else None,
        "item_count": len(g.items),
        "items": [_item_to_dict(it) for it in g.items],
    }


def _item_to_dict(it: models.ChartGroupItem) -> dict:
    if it.key_property_id and it.key_property:
        kp = it.key_property
        return {
            "id": it.id,
            "sort_order": it.sort_order,
            "source": "kp",
            "key_property_id": kp.id,
            "material": kp.material,
            "tc": kp.value_max,
            "pressure": kp.pressure_gpa,
            "type": kp.superconductor_type,
            "year": it.key_property.paper.year if it.key_property.paper else None,
            "doi": it.key_property.paper.doi if it.key_property.paper else None,
            "article_type": kp.article_type,
        }
    else:
        return {
            "id": it.id,
            "sort_order": it.sort_order,
            "source": "custom",
            "key_property_id": None,
            "material": it.custom_label,
            "tc": it.custom_tc,
            "pressure": it.custom_pressure,
            "type": it.custom_type,
            "year": it.custom_year,
            "doi": None,
            "article_type": it.custom_article_type,
        }


def _can_edit(g: models.ChartGroup, user: models.User) -> bool:
    if user.role == "superadmin":
        return True
    if user.role == "admin":
        return True
    if g.created_by == user.id:
        return True
    return False


def _can_set_public(user: models.User) -> bool:
    return user.role in ("admin", "superadmin")


def _can_delete(g: models.ChartGroup, user: models.User) -> bool:
    if user.role == "superadmin":
        return True
    if g.created_by == user.id:
        return True
    return False


class ItemIn(BaseModel):
    key_property_id: Optional[int] = None
    custom_label: Optional[str] = None
    custom_tc: Optional[float] = None
    custom_pressure: Optional[float] = None
    custom_type: Optional[str] = None
    custom_article_type: Optional[str] = None
    custom_year: Optional[int] = None


class GroupCreate(BaseModel):
    name: str
    description: Optional[str] = None
    items: list[ItemIn] = []


class GroupUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    items: Optional[list[ItemIn]] = None


# ── Routes ──

@router.get("", summary="列表（公开+预设+自己的）")
def list_groups(db: Session = Depends(get_db),
                current_user: models.User | None = Depends(get_current_user_optional)):
    if current_user:
        query = db.query(models.ChartGroup).filter(
            (models.ChartGroup.is_public == True) |
            (models.ChartGroup.is_preset == True) |
            (models.ChartGroup.created_by == current_user.id)
        )
    else:
        query = db.query(models.ChartGroup).filter(
            (models.ChartGroup.is_public == True) |
            (models.ChartGroup.is_preset == True)
        )
    groups = query.order_by(models.ChartGroup.is_preset.desc(), models.ChartGroup.updated_at.desc()).all()
    return [_group_to_dict(g) for g in groups]


@router.post("", summary="创建组合")
def create_group(body: GroupCreate, db: Session = Depends(get_db),
                 current_user: models.User = Depends(get_current_user)):
    g = models.ChartGroup(
        name=body.name, description=body.description,
        is_preset=False, is_public=False, created_by=current_user.id,
    )
    db.add(g)
    db.flush()
    for i, item in enumerate(body.items):
        db.add(models.ChartGroupItem(
            group_id=g.id, sort_order=i,
            key_property_id=item.key_property_id,
            custom_label=item.custom_label,
            custom_tc=item.custom_tc,
            custom_pressure=item.custom_pressure,
            custom_type=item.custom_type,
            custom_article_type=item.custom_article_type,
            custom_year=item.custom_year,
        ))
    db.commit()
    db.refresh(g)
    return _group_to_dict(g)


@router.post("/import", summary="从 JSON 导入组合")
def import_group(body: dict, db: Session = Depends(get_db),
                 current_user: models.User = Depends(get_current_user)):
    g = models.ChartGroup(
        name=body.get("name", "导入的组合"),
        description=body.get("description"),
        is_preset=False, is_public=False, created_by=current_user.id,
    )
    db.add(g)
    db.flush()
    for i, item in enumerate(body.get("items", [])):
        db.add(models.ChartGroupItem(
            group_id=g.id, sort_order=i,
            key_property_id=item.get("key_property_id") if item.get("type") == "kp" else None,
            custom_label=item.get("custom_label"),
            custom_tc=item.get("custom_tc"),
            custom_pressure=item.get("custom_pressure"),
            custom_type=item.get("custom_type"),
            custom_article_type=item.get("custom_article_type"),
            custom_year=item.get("custom_year"),
        ))
    db.commit()
    db.refresh(g)
    return _group_to_dict(g)


@router.get("/search", summary="搜索可加入组合的 key_properties")
def search_kps(q: str = Query(min_length=1), limit: int = 20,
               db: Session = Depends(get_db)):
    rows = db.query(models.KeyProperty).filter(
        models.KeyProperty.material.like(f"%{q}%")
    ).limit(limit).all()
    return [
        {
            "id": kp.id,
            "material": kp.material,
            "name": kp.name,
            "value_max": kp.value_max,
            "value_min": kp.value_min,
            "unit": kp.unit,
            "pressure_gpa": kp.pressure_gpa,
            "superconductor_type": kp.superconductor_type,
            "article_type": kp.article_type,
        }
        for kp in rows
    ]


@router.get("/{group_id}", summary="组合详情")
def get_group(group_id: int, db: Session = Depends(get_db),
              current_user: models.User = Depends(get_current_user)):
    g = db.query(models.ChartGroup).filter(models.ChartGroup.id == group_id).first()
    if not g:
        raise HTTPException(status_code=404, detail="组合不存在")
    return _group_to_dict(g)


@router.put("/{group_id}", summary="更新组合")
def update_group(group_id: int, body: GroupUpdate, db: Session = Depends(get_db),
                 current_user: models.User = Depends(get_current_user)):
    g = db.query(models.ChartGroup).filter(models.ChartGroup.id == group_id).first()
    if not g:
        raise HTTPException(status_code=404, detail="组合不存在")
    if not _can_edit(g, current_user):
        raise HTTPException(status_code=403, detail="无权编辑此组合")
    if body.name is not None:
        g.name = body.name
    if body.description is not None:
        g.description = body.description
    if body.items is not None:
        db.query(models.ChartGroupItem).filter(
            models.ChartGroupItem.group_id == group_id
        ).delete()
        for i, item in enumerate(body.items):
            db.add(models.ChartGroupItem(
                group_id=g.id, sort_order=i,
                key_property_id=item.key_property_id,
                custom_label=item.custom_label,
                custom_tc=item.custom_tc,
                custom_pressure=item.custom_pressure,
                custom_type=item.custom_type,
            custom_article_type=item.custom_article_type,
                custom_year=item.custom_year,
            ))
    g.updated_at = datetime.utcnow()
    db.commit()
    db.refresh(g)
    return _group_to_dict(g)


@router.delete("/{group_id}", summary="删除组合")
def delete_group(group_id: int, db: Session = Depends(get_db),
                 current_user: models.User = Depends(get_current_user)):
    g = db.query(models.ChartGroup).filter(models.ChartGroup.id == group_id).first()
    if not g:
        raise HTTPException(status_code=404, detail="组合不存在")
    if not _can_delete(g, current_user):
        raise HTTPException(status_code=403, detail="无权删除此组合")
    db.delete(g)
    db.commit()
    return {"message": "已删除"}


@router.post("/{group_id}/copy", summary="复制组合")
def copy_group(group_id: int, db: Session = Depends(get_db),
               current_user: models.User = Depends(get_current_user)):
    src = db.query(models.ChartGroup).filter(models.ChartGroup.id == group_id).first()
    if not src:
        raise HTTPException(status_code=404, detail="组合不存在")
    g = models.ChartGroup(
        name=f"{src.name} (副本)", description=src.description,
        is_preset=False, is_public=False, created_by=current_user.id,
    )
    db.add(g)
    db.flush()
    for item in src.items:
        db.add(models.ChartGroupItem(
            group_id=g.id, sort_order=item.sort_order,
            key_property_id=item.key_property_id,
            custom_label=item.custom_label,
            custom_tc=item.custom_tc,
            custom_pressure=item.custom_pressure,
            custom_type=item.custom_type,
            custom_article_type=item.custom_article_type,
            custom_year=item.custom_year,
        ))
    db.commit()
    db.refresh(g)
    return _group_to_dict(g)


@router.get("/{group_id}/export", summary="导出组合 JSON")
def export_group(group_id: int, db: Session = Depends(get_db)):
    g = db.query(models.ChartGroup).filter(models.ChartGroup.id == group_id).first()
    if not g:
        raise HTTPException(status_code=404, detail="组合不存在")
    return {
        "name": g.name,
        "description": g.description,
        "items": [
            {
                "type": "kp" if it.key_property_id else "custom",
                "key_property_id": it.key_property_id,
                "custom_label": it.custom_label,
                "custom_tc": it.custom_tc,
                "custom_pressure": it.custom_pressure,
                "custom_type": it.custom_type, "custom_article_type": it.custom_article_type,
                "custom_year": it.custom_year,
            }
            for it in g.items
        ],
    }


@router.patch("/{group_id}/public", summary="管理员切换公开状态")
def toggle_public(group_id: int, body: dict, db: Session = Depends(get_db),
                  current_user: models.User = Depends(get_current_user)):
    if not _can_set_public(current_user):
        raise HTTPException(status_code=403, detail="仅管理员可设置公开")
    g = db.query(models.ChartGroup).filter(models.ChartGroup.id == group_id).first()
    if not g:
        raise HTTPException(status_code=404, detail="组合不存在")
    g.is_public = body.get("is_public", not g.is_public)
    db.commit()
    return {"message": "已更新", "is_public": g.is_public}

File: goserver/handlers/chart_groups.go
package handlers

import (
	"net/http"
	"strconv"

	"scwiki/server/database"
	"scwiki/server/models"

	"github.com/gin-gonic/gin"
)

// ═══════════════════════════════════════════════
// 图表组合 API（替代 Python /api/chart-groups/*）
// ═══════════════════════════════════════════════

// ListChartGroups 列表（公开+预设+自己的）
// GET /api/chart-groups
func ListChartGroups(c *gin.Context) {
	var groups []models.ChartGroup
	query := database.DB.Preload("Items").Order("updated_at DESC")

	// 简单处理：返回全部公开+预设。认证用户额外看到自己的。
	isPublic := c.Query("is_public")
	isPreset := c.Query("is_preset")
	if isPublic == "true" {
		query = query.Where("is_public = ?", true)
	}
	if isPreset == "true" {
		query = query.Where("is_preset = ?", true)
	}

	query.Find(&groups)
	c.JSON(http.StatusOK, groups)
}

// GetChartGroup 组合详情
// GET /api/chart-groups/:id
func GetChartGroup(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))
	var group models.ChartGroup
	if err := database.DB.Preload("Items").First(&group, uint(id)).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "组合不存在"})
		return
	}
	c.JSON(http.StatusOK, group)
}

// CreateChartGroup 创建组合
// POST /api/chart-groups
func CreateChartGroup(c *gin.Context) {
	var body struct {
		Name        string `json:"name"`
		Description string `json:"description"`
		IsPublic    bool   `json:"is_public"`
		Items       []struct {
			KeyPropertyID     *uint    `json:"key_property_id"`
			CustomLabel       *string  `json:"custom_label"`
			CustomTc          *float64 `json:"custom_tc"`
			CustomPressure    *float64 `json:"custom_pressure"`
			CustomType        *string  `json:"custom_type"`
			CustomArticleType *string  `json:"custom_article_type"`
			CustomYear        *int     `json:"custom_year"`
		} `json:"items"`
	}
	if err := c.ShouldBindJSON(&body); err != nil || body.Name == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "名称不能为空"})
		return
	}

	group := models.ChartGroup{
		Name:    body.Name,
		IsPublic: body.IsPublic,
	}
	if body.Description != "" {
		group.Description = &body.Description
	}

	email, _ := c.Get("user_email")
	if email != nil {
		var user models.User
		if database.DB.Where("email = ?", email).First(&user).Error == nil {
			group.CreatedBy = &user.ID
		}
	}

	// SortOrder 用数组下标，保持前端传入顺序；与 UpdateChartGroup 一致。
	for i, item := range body.Items {
		it := models.ChartGroupItem{
			SortOrder:         i,
			KeyPropertyID:     item.KeyPropertyID,
			CustomLabel:       item.CustomLabel,
			CustomTc:          item.CustomTc,
			CustomPressure:    item.CustomPressure,
			CustomType:        item.CustomType,
			CustomArticleType: item.CustomArticleType,
			CustomYear:        item.CustomYear,
		}
		group.Items = append(group.Items, it)
	}

	database.DB.Create(&group)
	c.JSON(http.StatusOK, group)
}

// UpdateChartGroup 更新组合
// PUT /api/chart-groups/:id
func UpdateChartGroup(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))
	var group models.ChartGroup
	if err := database.DB.First(&group, uint(id)).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "组合不存在"})
		return
	}

	var body struct {
		Name        *string `json:"name"`
		Description *string `json:"description"`
		IsPublic    *bool   `json:"is_public"`
		Items       []struct {
			ID                *uint    `json:"id"`
			KeyPropertyID     *uint    `json:"key_property_id"`
			CustomLabel       *string  `json:"custom_label"`
			CustomTc          *float64 `json:"custom_tc"`
			CustomPressure    *float64 `json:"custom_pressure"`
			CustomType        *string  `json:"custom_type"`
			CustomArticleType *string  `json:"custom_article_type"`
			CustomYear        *int     `json:"custom_year"`
		} `json:"items"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "参数错误"})
		return
	}

	updates := map[string]any{}
	if body.Name != nil {
		updates["name"] = *body.Name
	}
	if body.Description != nil {
		updates["description"] = *body.Description
	}
	if body.IsPublic != nil {
		updates["is_public"] = *body.IsPublic
	}
	if len(updates) > 0 {
		database.DB.Model(&group).Updates(updates)
	}

	if body.Items != nil {
		database.DB.Where("group_id = ?", group.ID).Delete(&models.ChartGroupItem{})
		for i, item := range body.Items {
			it := models.ChartGroupItem{
				GroupID:           group.ID,
				SortOrder:         i,
				KeyPropertyID:     item.KeyPropertyID,
				CustomLabel:       item.CustomLabel,
				CustomTc:          item.CustomTc,
				CustomPressure:    item.CustomPressure,
				CustomType:        item.CustomType,
				CustomArticleType: item.CustomArticleType,
				CustomYear:        item.CustomYear,
			}
			database.DB.Create(&it)
		}
	}

	c.JSON(http.StatusOK, gin.H{"message": "已更新"})
}

// DeleteChartGroup 删除组合
// DELETE /api/chart-groups/:id
func DeleteChartGroup(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))
	database.DB.Where("group_id = ?", id).Delete(&models.ChartGroupItem{})
	database.DB.Delete(&models.ChartGroup{}, uint(id))
	c.JSON(http.StatusOK, gin.H{"message": "已删除"})
}

// TogglePublic 管理员切换公开状态
// PATCH /api/chart-groups/:id/public
func ToggleChartGroupPublic(c *gin.Context) {
	id, _ := strconv.Atoi(c.Param("id"))
	var body struct {
		IsPublic bool `json:"is_public"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "参数错误"})
		return
	}
	database.DB.Model(&models.ChartGroup{}).Where("id = ?", id).Update("is_public", body.IsPublic)
	c.JSON(http.StatusOK, gin.H{"message": "已更新"})
}
